<?xml version="1.0"?>
<launch>
    <!-- LAUNCH INTERFACE 
    This launches the hardware interface, loads the controllers and starts up a robot state publisher in the same namespace. 
    Use this as the default interface for the hardware, this file does not: load the joint names or the controller configuration. 
    The streamer argument will load the streamer controller, the robot will not be controlled using the fri but will output the joint state.
    
    example use:
        <include file="$(find iiwa_fri_ros)/launch/iiwa_hardware.launch">
            <arg name="name" value="iiwa"/>
            <arg name="hardware_interface" value="PositionJointInterface"/>
            <arg name="hostname" value="192.170.10.2"/>
            <arg name="port" value="30200"/>
            <arg name="controllers" value="joint_state_controller PositionJointInterface_trajectory_controller"/>
            <arg name="external_control" value="false"/>
        </include>           
    -->
    
    <arg name="name" default="iiwa"/>
    <arg name="hardware_interface" default="PositionJointInterface"/>
    <arg name="hostname" default="192.170.10.2"/>
    <arg name="port" default="30200"/>
    <arg name="controllers" default="joint_state_controller $(arg hardware_interface)_trajectory_controller"/>
    <arg name="external_control" default="false"/>
    <arg name="external_control_ip" default="172.31.1.149"/>
    
    <arg name="streamer" default="false"/>
    
    <group ns="$(arg name)">
        
        <!-- Load the controllers -->
        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
              output="screen" args="$(arg controllers) --shutdown-timeout 2" />
              
        <!-- Converts joint states to TF transforms for rviz, etc -->
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
              respawn="false" output="screen">  
        </node>

        <!-- Hardware interface to be loaded /-->
        <param name="interface" value="$(arg hardware_interface)"/>
        <param name="hostname" type="string" value="$(arg hostname)"/>
        <param name="port" value="$(arg port)"/>

        <!-- the real hardware interface /-->
        <group unless="$(arg streamer)">
            <node name="iiwa_fri_hardware" pkg="iiwa_fri_ros" type="iiwa_controller" respawn="false"/>
        </group>
        
        <group if="$(arg streamer)">
            <node name="iiwa_fri_hardware" pkg="iiwa_fri_ros" type="iiwa_streamer" respawn="false"/>
        </group>
        
        <!-- Launchs the external controller to start the kuka application -->
        <group if="$(arg external_control)">
            <node name="iiwa_external_start" pkg="iiwa_fri_ros" type="iiwa_app_start.py" args="$(arg external_control_ip)" respawn="false"/>
        </group>
        
    </group>
    
</launch>
